### PR TITLE
bush: prevent files being served outside of suite dir

### PR DESCRIPTION
**Pull request by oliver-sanders**
Prevent paths from being passed in url parameters (note file access is moderated by `os.access`).

### CODE REVIEW

**Comment by oliver-sanders** [comment]
* `t/rose-suite-restart/02-basic.t ` fixed by #2123 
* `t/syntax/00-lib.t` broken due to a change in version of the pep8 command which has now been renamed `pycodestyle` apparently.

File: lib/python/rose/bush.py [hunk 1]
```diff
@@ -741,13 +742,29 @@ def _get_user_suite_dir_root(self, user):
             self._get_user_home(user),
             self.bush_dao.SUITE_DIR_REL_ROOT))
 
+    @staticmethod
+    def _check_string_for_path(*strings):
+        for string in strings:
+            temp = os.path.split(string)
+            if len(temp) > 2 or (len(temp) == 2 and temp[0] != ''):
+                raise cherrypy.HTTPError(403)
```
Comment (matthewrmshin) [inline lib/python/rose/bush.py]: Suite name can look like a relative path, however.

Comment (oliver-sanders) [reply lib/python/rose/bush.py]: Good point.

File: lib/python/rose/bush.py [hunk 2]
```diff
@@ -741,13 +743,53 @@ def _get_user_suite_dir_root(self, user):
...
+    @staticmethod
+    def _check_string_for_path(string):
+        """Raise HTTP 403 error if the provided string contain path chars.
+
+        Raises:
+            cherrypy.HTTPError(403)
+
+        """
+        temp = os.path.split(string)
+        if len(temp) > 2 or (len(temp) == 2 and temp[0] != ''):
```
Comment (benfitzpatrick) [inline lib/python/rose/bush.py]: `temp` is always a pair according to docs? https://docs.python.org/2/library/os.path.html#os.path.split

**Comment by oliver-sanders** [comment]
`/t/syntax/00-lib.t` is fixed by #2128

**Comment by matthewrmshin** [comment]
Please add a test like this:

```diff
diff --git a/t/rose-bush/00-basic.t b/t/rose-bush/00-basic.t
index 5dc9df71..01bbd410 100755
--- a/t/rose-bush/00-basic.t
+++ b/t/rose-bush/00-basic.t
@@ -24,7 +24,7 @@ if ! python -c 'import cherrypy' 2>'/dev/null'; then
     skip_all '"cherrypy" not installed'
 fi
 
-tests 55
+tests 57
 
 ROSE_CONF_PATH= rose_ws_init 'rose' 'bush'
 if [[ -z "${TEST_ROSE_WS_PORT}" ]]; then
@@ -219,6 +219,15 @@ run_pass "${TEST_KEY}" \
     "${TEST_ROSE_WS_URL}/view/${USER}/${SUITE_NAME}?path=log/of/minus-one"
 file_grep "${TEST_KEY}.out" 'HTTP/.* 404 Not Found' "${TEST_KEY}.out"
 
+TEST_KEY="${TEST_KEY_BASE}-403-curl-view-outside"
+IMG_SUITE_DIR="$(mktemp -d --tmpdir="${HOME}/cylc-run" "rtb-rose-bush-00-XXXXXXXX")"
+echo 'Welcome to the imaginery suite.'>"${IMG_SUITE_DIR}/welcome.txt"
+run_pass "${TEST_KEY}" \
+    curl -I \
+    "${TEST_ROSE_WS_URL}/view/${USER}/${SUITE_NAME}?path=${IMG_SUITE_DIR}/welcome.txt"
+file_grep "${TEST_KEY}.out" 'HTTP/.* 403 Forbidden' "${TEST_KEY}.out"
+
 # test the file search feature
 TEST_KEY="${TEST_KEY_BASE}-200-curl-viewsearch"
 FILE='log/job/20000101T0000Z/foo1/01/job.out'
```

**Comment by matthewrmshin** [comment]
Tests broken by cylc/cylc#2475. I am rolling out a fix soon.

**Comment by matthewrmshin** [comment]
@arjclark do you mind giving this one a sanity check, please?

**Pull request merged**